pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/codestyle/UnnecessaryModifierRule.java

```diff
@@ -130,57 +211,75 @@ public Object visit(final ASTMethodDeclaration node, Object data) {
     
     public Object visit(final ASTResource node, final Object data) {
         if (node.isFinal()) {
-            addViolation(data, node);
+            reportUnnecessaryModifiers(data, node, Modifier.FINAL, "resources specifications are implicitly final");
         }
         
         return data;
     }
 
     public Object visit(ASTFieldDeclaration node, Object data) {
-        if (node.isSyntacticallyPublic() || node.isSyntacticallyStatic() || node.isSyntacticallyFinal()) {
-            check(node, data);
+        List<Modifier> unnecessary = new ArrayList<>();
+        if (node.isSyntacticallyPublic()) {
+            unnecessary.add(Modifier.PUBLIC);
+        }
+        if (node.isSyntacticallyStatic()) {
+            unnecessary.add(Modifier.STATIC);
         }
+        if (node.isSyntacticallyFinal()) {
+            unnecessary.add(Modifier.FINAL);
+        }
+
+        checkDeclarationInInterfaceType(data, node, unnecessary);
         return data;
     }
 
     public Object visit(ASTAnnotationMethodDeclaration node, Object data) {
-        if (node.isPublic() || node.isAbstract()) {
-            check(node, data);
+        List<Modifier> unnecessary = new ArrayList<>();
+        if (node.isPublic()) {
+            unnecessary.add(Modifier.PUBLIC);
+        }
+        if (node.isAbstract()) {
+            unnecessary.add(Modifier.ABSTRACT);
         }
+        checkDeclarationInInterfaceType(data, node, unnecessary);
         return data;
     }
     
     public Object visit(ASTConstructorDeclaration node, Object data) {
         if (node.getNthParent(2) instanceof ASTEnumBody) {
             if (node.isPrivate()) {
-                addViolation(data, node);
+                reportUnnecessaryModifiers(data, node, Modifier.PRIVATE, "enum constructors are implicitly private");
             }
         }
         return data;
     }
-    
+
+
     private boolean isSafeVarargs(final ASTMethodDeclaration node) {
-        for (final ASTAnnotation annotation : node.jjtGetParent().findChildrenOfType(ASTAnnotation.class)) {
-            final Node childAnnotation = annotation.jjtGetChild(0);
-            if (childAnnotation instanceof ASTMarkerAnnotation) {
-                final ASTMarkerAnnotation marker = (ASTMarkerAnnotation) childAnnotation;
-                if (marker.getType() != null && SafeVarargs.class.isAssignableFrom(marker.getType())) {
-                    return true;
-                }
-            }
-        }
-        
-        return false;
+        return node.isAnnotationPresent(SafeVarargs.class.getName());
     }
 
-    private void check(Node fieldOrMethod, Object data) {
+
+    private void checkDeclarationInInterfaceType(Object data, Node fieldOrMethod, List<Modifier> unnecessary) {
```

Comment:
possibly better to use an `EnumSet<Modifier>` both for performance and avoiding duplicates